### ISSUE TITLE
Apply a JSON workflow on a subset of rows with facets

### ISSUE THREAD

**Issue by wetneb** [OpenRefine/OpenRefine#2881]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**

I have some facets which select a subset of rows, and a list of operations represented as a JSON workflow. If I use the history tab to apply this JSON workflow to the project, it will apply to all rows, not just the ones I have selected with my facets.

**Describe the solution you'd like**

It should be possible to apply a given recipe to a filtered set of rows determined by facets. Not all operations respect facets, so this would only be useful if the workflow only contains operations which do respect them.

**Describe alternatives you've considered**

At the moment, the only solution I can think of is to edit the JSON payload manually to add the facet configurations in the `engineConfig` parameters of each operation, which is not convenient.

**Comment by thadguidry** [comment]
I thought we had this issue somewhere before? hmm...

**Comment by tfmorris** [comment]
> If I use the history tab to apply this JSON workflow to the project, it will apply to all rows, not just the ones I have selected with my facets.

I'm surprised at that. I would have expected it to use the facet settings from when the operation history was created since the facet settings are an intrinsic part of the original operations. BTW, if you think of the operation undo history as a "workflow", you are bound to experience ongoing cognitive dissonance because that's not what it was designed to be at all.

**Comment by wetneb** [comment]
I would be curious to know what it was designed for, actually! I cannot think of another use for this dialog :)

**Comment by thadguidry** [comment]
The original use case came from Wikipedia table extraction (sets of tables for Olympics, Sports, etc.) and re-applying the same set of transformations on all of them to then upload to Freebase.  It was... "a very basic intent" of a repeatable workflow.

**Comment by wetneb** [comment]
Ah nice! I hope that if there are still people doing that out there, they find the Wikitext importer useful… https://www.wikidata.org/wiki/Wikidata:Tools/OpenRefine/Editing/Tutorials/Inverse_Listeria

**Comment by thadguidry** [comment]
@wetneb I'm sure there are.

**Comment by ostephens** [comment]
>I have some facets which select a subset of rows, and a list of operations represented as a JSON workflow. If I use the history tab to apply this JSON workflow to the project, it will apply to all rows, not just the ones I have selected with my facets.

I'm sure this used to work - the operation history captures the facets and selections afaik? Can we look at a concrete example?

**Comment by wetneb** [comment]
> the operation history captures the facets and selections afaik

Yes it does: it captures the facets and selections which were on when the operation was originally applied.
But then, when you apply an operation history to a new project, the facets you have while you paste the operation history and apply it will not be taken into account.

Example:
- I create a text filter and use it to select the cells containing "foo"
- I apply a row-wise operation (say a mass edit)
- I extract the operation in the undo/redo tab. I obtain:
```
[
  {
    "op": "core/mass-edit",
    "engineConfig": {
      "facets": [
        {
          "type": "text",
          "name": "a",
          "columnName": "a",
          "query": "foo",
          "mode": "text",
          "caseSensitive": false,
          "invert": false
        }
      ],
      "mode": "row-based"
    },
    "columnName": "b",
    "expression": "value",
    "edits": [
      {
        "from": [
          "hello"
        ],
        "fromBlank": false,
        "fromError": false,
        "to": "hi"
      }
    ],
    "description": "Mass edit cells in column b"
  }
]
```
* on a different project, I create a numeric facet to select values above 42 in some column
* while that facet is filtering out rows, I go to the history tab and paste the JSON above.

At the moment, the numeric facet I have created will be ignored: the mass edit will only be restricted by the text filter.
In other words, if I extract the operation's JSON representation afterwards, I will still have the exact same thing as above - there will not be a second facet in there.

**Comment by ostephens** [comment]
OK - I understand now.

So this is about respecting additional facets not specified in the original set of operations.
This seems quite complex - I think the extract/apply operations history functions would need to be more sophisticated than currently to allow you to decide what facets were/were not applied at time of extraction and application?

**Comment by wetneb** [comment]
In terms of architecture it seems doable to me: we have a common base class for all operations which respect facets ("engine dependent" operations in the code), which exposes the facets they had at the time ("engine config"). So it is essentially a matter of adding the current facets to those engine configs after parsing the list of operations.

But it is not clear to me whether these additional facets should be applied by default, or whether we should give the user the option to keep the current behaviour.

**Comment by wetneb** [comment]
(that being said I am not really thinking about fixing this in the current UI as things stand, this is more of a way to document the current drawbacks of the existing system, with the hope to design a proper solution for reproducibility - perhaps the mailing list is a better place to do that)

**Comment by ostephens** [comment]
>But it is not clear to me whether these additional facets should be applied by default, or whether we should give the user the option to keep the current behaviour.

This needs to be user controlled IMO

**Comment by tfmorris** [comment]
@wetneb Can you update the description to match the understanding arrived at (ie it's not "all rows", but the original facet selections)?

We should definitely warn users that any facet settings will have no effect when an operation history is replayed and that the original facet settings are being used.

**Comment by wetneb** [comment]
@tfmorris hopefully the new wording is clearer, feel free to improve if you see a better way to phrase that

**Comment by tfmorris** [comment]
@wetneb The main thing I was concerned about was the "all rows" reference. I've tweaked it slightly to remove that. Please make sure it's still technically accurate. After all this discussion I'm confused myself and don't have time to go and check the code immediately.

**Comment by wetneb** [comment]
Yes you got it right, removing the striked out "all" to make it clearer for others.
